contracts/payment/RefundTokenEscrow.sol

```diff
@@ -0,0 +1,78 @@
+pragma solidity ^0.4.24;
+
+import "./ConditionalTokenEscrow.sol";
+import "../token/ERC20/ERC20.sol";
+
+
+/**
+ * @title RefundTokenEscrow
+ * @dev Escrow that holds tokens for a beneficiary, deposited from multiple parties.
+ * The contract owner may close the deposit period, and allow for either withdrawal
+ * by the beneficiary, or refunds to the depositors.
+ */
+contract RefundTokenEscrow is ConditionalTokenEscrow {
+
+  enum State { Active, Refunding, Closed }
+
+  event Closed();
+  event RefundsEnabled();
+
+  State public state;
+  address public beneficiary;
+
+  /**
+   * @dev Constructor.
+   * @param _token Address of the ERC20 token that will be put in escrow.
+   * @param _beneficiary The beneficiary of the deposits.
+   */
+  constructor(ERC20 _token, address _beneficiary) public TokenEscrow(_token) {
+    require(_beneficiary != address(0));
+    beneficiary = _beneficiary;
+    state = State.Active;
+  }
+
+  /**
+   * @dev Stores tokens that may later be refunded.
+   * @param _refundee The address tokens will be sent to if a refund occurs.
+   * @param _amount The amount of tokens to store.
+   */
+  function deposit(address _refundee, uint256 _amount) public {
+    require(state == State.Active);
+    super.deposit(_refundee, _amount);
```

Comment:
Let me see if get your point.
**As it is**, in pseudocode, a simplified happy path would be as follows.
_Note that `owner` always refers to the owner of `escrow` and Payer is the same as `refundee`_
1. Payer calls `token.approve(owner, N)` to allow `owner` transfer N tokens
2. Payer tells the `owner` she wants to deposit N tokens to `beneficiary`
3. `owner` calls `token.transferFrom(payer, owner, N)` to get hold of N tokens (transfer was approved in 1)
4. `owner` approves `escrow` to move N tokens and calls `escrow.deposit(beneficiary, N)` to put N tokens in escrow
5. `escrow` calls `token.transferFrom(owner, escrow, N)` to get hold of N tokens (transfer was approved in 4).
6. Supposing all conditions are met, `beneficiary` tells the `owner` she wants to withdraw the tokens.
7. `owner` calls `escrow.withdraw(beneficiary)`
8. `escrow` calls `token.transfer(beneficiary, N)`

If I understand, **you suggest** to do something like this:
1. Payer calls `token.approve(escrow, N)` to allow `escrow` transfer N tokens
2. Payer tells the `owner` she wants to deposit N tokens to `beneficiary`
3. `owner` calls `escrow.deposit(beneficiary, N)` to put N tokens in escrow
4. `escrow` calls `token.transferFrom(payer, escrow, N)` to get hold of N tokens (transfer was approved in 1).
5. Supposing all conditions are met, `beneficiary` tells the `owner` she wants to withdraw the tokens.
6. `owner` calls `escrow.withdraw(beneficiary)`
7. `escrow` calls `token.transfer(beneficiary, N)`

My concerns about this last flow:
> 1. Payer calls `token.approve(escrow, N)` to allow `escrow` transfer N tokens

 I'm not sure about this step in terms of 'payer experience'. First, this would force the payer to know the address of the `escrow` in addition to the address of the `owner`. Furthermore, it would also require the payer to exactly understand how the payment method works behind the scenes (otherwise, the question would be: who's `escrow`? Why do I have to allow him to take my tokens if I just wanted to deposit to `beneficiary` through `owner`?). Some abstraction from how the payment method is actually implemented might be better for certain payers.

>4. `escrow` calls `token.transferFrom(payer, escrow, N)` to get hold of N tokens (transfer was approved in 1).

How would `escrow` know the payer's address in this case? Should it be added as a param to the `deposit` function in step 3? In that case, we would be adding yet another difference between `TokenEscrow` and `Escrow`.

Reply:
Identifying potential differences between both escrows is exactly what I intend to do :)

I agree with your assessment however: exposing the escrow address, while bad, was not _too_ bad, but having to tell the escrow who to transfer from (or simply using the refundee as this address) feels way off. I'm still not thrilled with two transfers taking place when only one is needed, but I guess it'll do for now, I can't think of a better way of doing this.

Note btw that, during step 4 in the first scenario, `owner` need not `approve` the `escrow`, since this can be done at construction time for an amount of `MAX_UINT256`.